### PR TITLE
Addition of Profile Configuration CLI Option

### LINKED ISSUE #6323: Creation of Configuration Profiles
## Description
As a user, I want a `--profile` flag that I can set to easily overwrite some defaults. This way, as a new user I can only set one flag and have a good experience out of the box and get familiar with operating Besu more quickly.

### Acceptance Criteria
- New configuration option `--profile=string` 
- This flag, when set, overrides the defaults in the client with new ones from the parsed profiles.
- The profiles flag accepts the following profiles (expanded on in other tickets):
    - selfish-staker #6324
    - staker #6325
    - private network && enterprise (both should alias to the same profile) #6326
 - Any configuration options explicitly set in the config file or command line should overwrite any of the defaults set via the profile (i.e. private network defaults to `sync-mode=FAST`, if a user sets `sync-mode=FULL`, then the sync mode should be full

### PR DESCRIPTION
**Pull request by Gabriel-Trintinalia**
## PR description
This pull request introduces the `--profile` CLI option to Besu. This new feature allows users to load existing TOML configuration files from resources, enabling the override of default options. This PR does not create the profiles - they will be created in different PRs.

- Profile CLI Option: The `--profile` CLI option has been added, providing users the ability to set the profile Besu should utilise for its configuration.

- Configuration Resolution: The configuration provider resolves settings in a cascading manner, prioritising environment variables, followed by the configuration file, the profile file (if any), and finally falling back to the variable default if no other configuration source is available.

For instance, consider a scenario where you have two TOML configuration files: `config.toml` and `staker.toml`. The `config.toml` file is provided by the user via the `--config-file` option, and `staker.toml` is a pre-configured resource file containing custom settings. The staker.toml file can be specified using the new --profile option:

`besu --config-file=config.toml --profile=STAKER`

In this case, Besu would first look for settings in the environment variables. If a setting is not found in the environment variables, Besu would look for it in the config.toml file. If the setting is not found in the config.toml file, Besu would then look for it in the `staker.toml` file. If the setting is not found in the `staker.toml` file, Besu would use the default value for that setting.

The rationale for using the TOML format to define profiles is as follows:

- It leverages existing TOML parsing and validation mechanisms
- The format is user-friendly
- It consolidates all configuration data into a single location

## Fixed Issue(s)
fixes #6323

### CODE REVIEW

**Comment by Gabriel-Trintinalia** [comment]
related #6409

File: besu/src/test/java/org/hyperledger/besu/cli/CascadingDefaultProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+          .put("config", (new JsonObject()).put("chainId", GENESIS_CONFIG_TEST_CHAINID));
+
+  /**
+   * Test if the default values are overridden if the key is present in the configuration file. The
+   * test checks if the configuration file correctly overrides the default values for various
+   * settings, such as the JSON-RPC configuration, GraphQL configuration, WebSocket configuration,
+   * and metrics configuration.
+   */
+  @Test
+  public void overrideDefaultValuesIfKeyIsPresentInConfigFile(final @TempDir File dataFolder)
```
Comment (pinges) [inline besu/src/test/java/org/hyperledger/besu/cli/CascadingDefaultProviderTest.java:71]: I do not really understand how this test is testing what is stated in the comment. We should talk :-)

File: besu/src/test/resources/everything_config.toml [hunk 2]
```diff
@@ -16,7 +16,7 @@ node-private-key-file="./path/to/privateKey"
 pid-path="~/.pid"
 reorg-logging-threshold=0
 static-nodes-file="~/besudata/static-nodes.json"
-
+profile="NONE"
```
Comment (pinges) [inline besu/src/test/resources/everything_config.toml:19]: The profile NONE does not exist in ProfileName. Should that not cause an error when the everything_config.toml is used? (e.g. in BesuCommandTest)

File: besu/src/main/java/org/hyperledger/besu/cli/util/TomlConfigurationDefaultProvider.java [hunk 3]
```diff
@@ -70,35 +104,32 @@ public String defaultValue(final ArgSpec argSpec) {
...
+        || type.equals(Long.class)
+        || type.equals(long.class)
+        || type.equals(Wei.class)
+        || type.equals(BigInteger.class)
+        || type.equals(Double.class)
+        || type.equals(double.class)
+        || type.equals(Float.class)
+        || type.equals(float.class)
+        || type.equals(Percentage.class)
+        || type.equals(Fraction.class);
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/util/TomlConfigurationDefaultProvider.java:132]: nice

File: besu/src/test/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilderTest.java [hunk 4]
```diff
@@ -209,4 +210,11 @@ void setWorldStateUpdateModeJournaled() {
     final String layeredTxPoolSelected = builder.build();
     assertThat(layeredTxPoolSelected).contains("Using JOURNALED worldstate update mode");
   }
+
+  @Test
+  void setProfile() {
+    builder.setProfile(ProfileName.DEV.name());
+    final String layeredTxPoolSelected = builder.build();
```
Comment (macfarla) [inline besu/src/test/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilderTest.java:217]: layeredTxPoolSelected seems an unrelated name for this variable

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 5]
```diff
@@ -536,6 +537,13 @@ private InetAddress autoDiscoverDefaultIP() {
               + " (default: ${DEFAULT-VALUE})")
   private final NetworkName network = null;
 
+  @Option(
+      names = {PROFILE_OPTION_NAME},
+      paramLabel = PROFILE_FORMAT_HELP,
+      description =
+          "Overwrite default settings. Possible values are ${COMPLETION-CANDIDATES}. (default: none)")
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:544]: Not sure what this should say exactly... Just created #6430 to add something in the help footer so it shows at the bottom, so I think we can refine this later. 

Choose a profile to run Besu with that set of defaults. For more info see...

File: besu/src/main/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilder.java [hunk 6]
```diff
@@ -273,7 +285,8 @@ public String build() {
     final List<String> lines = new ArrayList<>();
     lines.add("Besu version " + BesuInfo.class.getPackage().getImplementationVersion());
     lines.add("");
-    lines.add("Configuration:");
+    String profileInfo = (profile != null) ? String.format(" %s", profile) : "";
+    lines.add(String.format("Configuration%s:", profileInfo));
```
[the remark this answers is not in the document]
Comment (Gabriel-Trintinalia) [reply besu/src/main/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilder.java]: if no profile is selected, profile section does not appear

File: besu/src/test/resources/everything_config.toml (hunk 2 above)
Comment (Gabriel-Trintinalia) [reply besu/src/test/resources/everything_config.toml:19]: that does not cause an error, in fact that appears several times in the `everything_config`

File: besu/src/main/java/org/hyperledger/besu/cli/util/ConfigFileFinder.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  @Override
+  public Optional<File> getFromOption(
+      final CommandLine.ParseResult parseResult, final CommandLine commandLine) {
+    final CommandLine.Model.OptionSpec configFileOption =
+        parseResult.matchedOption(CONFIG_FILE_OPTION_NAME);
+    try {
+      File file = configFileOption.getter().get();
+      if (!file.exists()) {
+        throw new CommandLine.ParameterException(
+            commandLine, "Unable to read TOML configuration, file not found.");
```
[the remark this answers is not in the document]
Comment (Gabriel-Trintinalia) [reply besu/src/main/java/org/hyperledger/besu/cli/util/ConfigFileFinder.java]: done

File: besu/src/test/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilderTest.java (hunk 4 above)
Comment (Gabriel-Trintinalia) [reply besu/src/test/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilderTest.java]: Fixed

**Review by macfarla — APPROVED** [review]
LGTM
